File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        ctp = {']':'[','}':'{',')':'('}

        for i in range(len(s)):
            if s[i] in ctp:
                if stack and stack[-1] == ctp[s[i]]:
                    stack.pop()
                else:
                    return False
            
            else:
                stack.append(s[i])
        
        return len(stack) == 0
        

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
       Stack<Character> stack = new Stack<>();
        HashMap<Character,Character> ctp = new HashMap<>();

        ctp.put('}','{');
        ctp.put(']','[');
        ctp.put(')','(');

       for (int i = 0; i < s.length(); i++)
       {
           if(ctp.containsKey(s.charAt(i))){
               if(!stack.isEmpty() && stack.lastElement() == ctp.get(s.charAt(i)))
               {
                   stack.pop();
               }
               else
                   return false;
           }
           else
               stack.add(s.charAt(i));
       }
       return stack.isEmpty();
    }
}